### PR TITLE
Lesson25

### PR DESCRIPTION
**Pull request by thelast27**
Check my changes, pls 
This commit was created for example matters, but all project is fresh. u may check _View Controller_ and _WeatherParameters_ files.

### CODE REVIEW

**Comment by vkhadyka** [comment]
https://github.com/thelast27/weather_app_Eldar/blob/23b9343dcdcc2ca6b87d2720aa98a26a3d413f89/weatherApp/weatherApp/ViewController.swift#L17

move out of string `appid`

**Comment by vkhadyka** [comment]
> without `appid` doesnt works sever communication. Or u didt mean it? 

I meant that you should store appid value to constant, for example, or somewhere else

**Comment by thelast27** [comment]
I added my API to info plist, but I don't know how to read it from. Can you explain at lesson today?

**Comment by vkhadyka** [comment]
Still have an issue with that? Let figure out it and merge PR.

File: weatherApp/weatherApp/ViewController.swift [hunk 1]
```diff
@@ -8,32 +8,50 @@
...
-          let dataTaks = URLSession.shared.dataTask(with: urlRequest) { data, response, error in
-
-              if let data = data {
-                  guard let weatherParametrs = try? JSONDecoder().decode(ResponseBody.self, from: data) else { return }
-                  print("In \(weatherParametrs.name) today \(weatherParametrs.main.temp) by Celsius")
-              }
-          }
-          dataTaks.resume()
+        weatherManager.urlRequest(city: "Dublin")
+        weatherManager.completion = { [weak  self] currentWeather in
```
Comment (vkhadyka) [inline weatherApp/weatherApp/ViewController.swift:25]: why did do decide implement completion by that way?

File: weatherApp/weatherApp/WeatherManager .swift [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        case english
+        case arabic
+        case german
+        case french
+        case japanese
+        case chinese
+        case italian
+        case ukrainian
+        
+        var shortName: String {
```
Comment (vkhadyka) [inline weatherApp/weatherApp/WeatherManager .swift:28]: I've seen it before. I'll try to realize where...

File: weatherApp/weatherApp/WeatherManager .swift [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                } catch let error {
+                    print(error)
+                }
+            }
+        }
+        dataTask.resume()
+    }
+    
+    func currentWeather(long: Double, lat: Double, withLang lang: Languages) {
+        let urlAddress = "https://api.openweathermap.org/data/2.5/onecall?lat=\(lat)&lon=\(long)&exclude=daily&appid=\(weatherApiKey)&lang=\(lang.shortName)&units=metric"
```
Comment (vkhadyka) [inline weatherApp/weatherApp/WeatherManager .swift:73]: baseURL move out of string.
`units` is hardcoded.

File: weatherApp/weatherApp/WeatherManager .swift [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        guard let url = URL(string: urlAddress) else { return }
+        var urlRequest = URLRequest(url: url)
+        urlRequest.httpMethod = "GET"
+        let session = URLSession(configuration: .default)
+        let dataTask = session.dataTask(with: urlRequest) { [weak self] data, response, error in
+            if let error = error {
```
Comment (vkhadyka) [inline weatherApp/weatherApp/WeatherManager .swift:79]: why do you extract optional `self` in line 52, but missed here?

File: weatherApp/weatherApp/WeatherParameters.swift [hunk 5]
```diff
@@ -36,7 +36,7 @@ struct ResponseBody: Decodable {
     }
     
 }
-//можно и как енам. Просто нашел такой способ - решил показать 
+
 extension ResponseBody.MainResponse {
```
[the remark this answers is not in the document]
Comment (thelast27) [reply weatherApp/weatherApp/WeatherParameters.swift]: Found in google. We can rename vars via extantion or enum. Enum mow useful.

File: weatherApp/weatherApp/ViewController.swift (hunk 1 above)
Comment (thelast27) [reply weatherApp/weatherApp/ViewController.swift:25]: for avoiding memory leak

File: weatherApp/weatherApp/WeatherManager .swift (hunk 3 above)
Comment (thelast27) [reply weatherApp/weatherApp/WeatherManager .swift:73]: need also put it in some enum or smth?

File: weatherApp/weatherApp/ViewController.swift (hunk 1 above)
Comment (vkhadyka) [reply weatherApp/weatherApp/ViewController.swift:25]: I meant why do you implement closure out of method?

File: weatherApp/weatherApp/WeatherManager .swift (hunk 3 above)
Comment (vkhadyka) [reply weatherApp/weatherApp/WeatherManager .swift:73]: smth, yep

File: weatherApp/weatherApp/WeatherManager .swift (hunk 4 above)
Comment (thelast27) [reply weatherApp/weatherApp/WeatherManager .swift:79]: add self unwrapping between 78 and 79 lines

Comment (vkhadyka) [reply weatherApp/weatherApp/WeatherManager .swift:79]: не вижу коммита

File: weatherApp/weatherApp/ViewController.swift [hunk 6]
```diff
@@ -53,5 +54,16 @@ class ViewController: UIViewController {
             self.sunIndexLable.text = "Sun Index is \(Double(sunIndex))"
         }
     }
-    
+//    MARK: - set background image with blure effect
+    func setbackgroundPic() {
+            let backgroundImage = UIImageView(frame: UIScreen.main.bounds)
```
Comment (vkhadyka) [inline weatherApp/weatherApp/ViewController.swift:59]: у меня тут 2 вопроса.
1. Почему не сделал в storyboard?
2. Почему не задавал размеры констрейнтами?

File: weatherApp/weatherApp/WeatherManager .swift [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            case .italian: return "it"
+            case .ukrainian: return "ua"
+            }
+        }
+    }
+    
+    var completion: ((CurrentAndForecastWeather) -> Void)?
+    
+    func urlRequest(city: String) {
+        let urlAddress = "https://api.openweathermap.org/geo/1.0/direct?q=\(city)&appid=\(weatherApiKey)"
```
[the remark this answers is not in the document]
Comment (thelast27) [reply weatherApp/weatherApp/WeatherManager .swift]: fixed

**Review by vkhadyka — APPROVED** [review]

File: weatherApp/weatherApp/ViewController.swift (hunk 6 above)
Comment (thelast27) [reply weatherApp/weatherApp/ViewController.swift:59]: 2. fixed

**Review by vkhadyka — APPROVED** [review]

File: weatherApp/weatherApp/ViewController.swift (hunk 6 above)
Comment (vkhadyka) [reply weatherApp/weatherApp/ViewController.swift:59]: так нет ж, твой `backgroundImage` не привязан констрейнтами к superview

Comment (thelast27) [reply weatherApp/weatherApp/ViewController.swift:59]: да, два случайно привязались к safe area 
поправил

Comment (thelast27) [reply weatherApp/weatherApp/ViewController.swift]: убрал создание блюд вью кодом, перенес всё это на сториборд, добавил констрейнты к супер вью. Прочертил и на СЕ первого поколения и на 12 дюймовом айпэде.

File: weatherApp/weatherApp/WeatherManager .swift [hunk 8]
```diff
@@ -36,10 +36,18 @@ class WeatherManager {
...
     
     var completion: ((CurrentAndForecastWeather) -> Void)?
-    
+// MARK: - get geocoding data
     func urlRequest(city: String) {
-        let urlAddress = "https://api.openweathermap.org/geo/1.0/direct?q=\(city)&appid=\(weatherApiKey)"
-        guard let url = URL(string: urlAddress) else { return }
+        var components = URLComponents()
+        components.scheme = "https"
+        components.host = "api.openweathermap.org"
```
Comment (vkhadyka) [inline weatherApp/weatherApp/WeatherManager .swift:43]: тут не мешало бы вынести все-таки урл.

Comment (thelast27) [reply weatherApp/weatherApp/WeatherManager .swift:43]: done

File: weatherApp/weatherApp/ViewController.swift [hunk 9]
```diff
@@ -9,26 +9,31 @@ import UIKit
 
 class ViewController: UIViewController {
     
+    var topConstraint = NSLayoutConstraint()
```
[the remark this answers is not in the document]
Comment (thelast27) [reply weatherApp/weatherApp/ViewController.swift:12]: done

**Pull request merged**